
### `r region_name`

#### Fine mapping

Fine mapping was performed by evaluating the genotype-phenotype relationship for variants nearby the QTL identified from GWA mapping using a vcf containing imputed variants to avoid removing variants with missing genotype information for one or a few strains. Only SNVs were considered in this mapping.

Each variant is represented by a vertical line, colored by the predicted variant impact (i.e. HIGH impact variants could be variants that introduce a change in the amino acid sequence or a stop-gain). Genes are represented by horizontal lines with an arrow showing the direction of the gene. 

```{r, out.width = '100%', fig.height = 6}
# read in genes in region and pval
genes_in_region <- data.table::fread(glue::glue("../../{alg}/Fine_Mappings/Data/{trait_name}_{QTL_chrom}_{QTL_start}-{QTL_end}_bcsq_genes_{stringr::str_to_lower(alg)}.tsv"))

gene_df <- genes_in_region %>%
    dplyr::filter(START_POS == unique(genes_in_region$START_POS)) %>%
    dplyr::distinct(WBGeneID, GENE_NAME, PEAK_MARKER, CHROM, STRAND, TRANSCRIPTION_START_POS, 
                    TRANSCRIPTION_END_POS, START_POS, END_POS, VARIANT_LOG10p) %>%
    dplyr::group_by(WBGeneID) %>%
    dplyr::mutate(VARIANT_LOG10p = max(VARIANT_LOG10p, na.rm = T)) %>%
    dplyr::distinct()

peak_variant <- as.numeric(strsplit(unique(gene_df$PEAK_MARKER), split = ":")[[1]][2])

variant_df <- genes_in_region %>%
    dplyr::filter(START_POS == unique(genes_in_region$START_POS)) %>%
    dplyr::distinct(CHROM, POS, GENE_NAME, WBGeneID, VARIANT_LOG10p, VARIANT_IMPACT)

variant_df$VARIANT_IMPACT[is.na(variant_df$VARIANT_IMPACT)] <- "Intergenic"

xs <- unique(gene_df$START_POS)
xe <- unique(gene_df$END_POS)
cq <- unique(gene_df$CHROM)

max_logp <- unique(max(variant_df$VARIANT_LOG10p, na.rm = T))/150

gene_plot <- ggplot(gene_df) +
    aes(text = paste0(GENE_NAME, "\n", WBGeneID)) +
    geom_vline(aes(xintercept = peak_variant/1e6),
               linetype=3, color = "cyan")+
    geom_segment(aes(x = ifelse(STRAND == "+", TRANSCRIPTION_START_POS/1e6, TRANSCRIPTION_END_POS/1e6),
                     xend = ifelse(STRAND == "+", TRANSCRIPTION_END_POS/1e6, TRANSCRIPTION_START_POS/1e6),
                     y = VARIANT_LOG10p,
                     yend = VARIANT_LOG10p),
                 arrow = arrow(length = unit(5, "points")), size = 1) +
    geom_segment(aes(x = POS/1e6,
                     xend = POS/1e6,
                     y = VARIANT_LOG10p+max_logp,
                     yend = VARIANT_LOG10p-max_logp,
                     color = VARIANT_IMPACT), data = variant_df) +
    # if snpeff, need to add back moderate and modifier
    scale_color_manual(values = c("LOW" = "gray30",
                                  "HIGH" = "red",
                                  "Linker" = "gray80", 
                                  "Intergenic" = "gray80"),
                       breaks = c("HIGH", "LOW", "Intergenic"),
                       name = "EFFECT")+
    labs(x = "Genomic Position (Mb)",
         y = "-log10p") +
         # y = expression(-log[10](italic(p))))+
    theme_bw(18)+
    xlim(c(xs/1e6, xe/1e6)) +
    theme(legend.position = "top",
          panel.grid = element_blank())

# print(gene_plot)

# plotly::ggplotly(gene_plot, tooltip = c("plotly_label_gene_name", "plotly_label_gene_ID", "plotly_label_variant")) %>% plotly::toWebGL()

# if finemap region is too large, don't plot plotly and don't plot next finemap plot
if(QTL_end - QTL_start < 5e6) {
  # plotly::ggplotly(gene_plot, tooltip = "text")
  print(gene_plot)
  finemap2 <- TRUE
} else {
  print("Region too large for interactive plot:")
  print(gene_plot)
  finemap2 <- FALSE
}

    
    # ggsave(gene_plot,
    #        filename = glue::glue("{analysis_trait}_{cq}_{xs}-{xe}_gene_plot_{ann_type}.pdf"),
    #        height=10, width = 14)



```

```{r, out.width = '100%', fig.height = 6, eval = finemap2}

# skip this for now, plot not necessary and it takes up lots of memory running
cat("\n This second plot is very similar to the first. Here, each variant is represented by a diamond colored by the linkage to the peak marker (colored in red). This plot can be useful to determine what the strucutre of your region looks like. If you have many variants with high linkage to your peak marker, it is important to remember that any of those variants could be causal.")

finemap <- data.table::fread(glue::glue("../../{alg}/Fine_Mappings/Data/{trait_name}.{QTL_chrom}.{QTL_start}.{QTL_end}.finemap_inbred.{stringr::str_to_lower(alg)}.fastGWA")) %>%
  dplyr::mutate(CHR = as.factor(CHR),
                SNP = as.factor(SNP),
                POS = as.numeric(POS),
                AF1 = as.numeric(AF1),
                P = as.numeric(P))
ROI.geno.matrix <- data.table::fread(glue::glue("../../{alg}/Fine_Mappings/Data/{trait_name}.{QTL_chrom}:{QTL_start}-{QTL_end}.ROI_Genotype_Matrix_{stringr::str_to_lower(alg)}.tsv"))
ROI.LD <- data.table::fread(glue::glue("../../{alg}/Fine_Mappings/Data/{trait_name}.{QTL_chrom}.{QTL_start}.{QTL_end}.LD_{stringr::str_to_lower(alg)}.tsv"))


peakp <- unique(ROI.LD$BP_A)

finemap_peaks <- na.omit(finemap) %>%
  dplyr::filter(POS == peakp)

finemap_peaks$POS <- as.numeric(finemap_peaks$POS)

# add strain genotype to ld file for snpeff later
roi_genotype <- ROI.geno.matrix %>%
  tidyr::gather(strain, allele, -c(CHROM:ALT)) %>%
  dplyr::mutate(allele = ifelse(allele == -1, "REF", ifelse(allele == 1, "ALT", NA))) %>%
  dplyr::group_by(CHROM, POS, REF, ALT, allele) %>%
  dplyr::summarize(strains = paste(strain, collapse = ",")) 

pr_roi_ld <- ROI.LD %>%
  dplyr::mutate(peak_marker = gsub("_", ":", unique(finemap_peaks$SNP))) %>%
  dplyr::mutate(SNP = SNP_B) %>%
  dplyr::select(peak_marker, peak_maf = MAF_A, SNP, maf_marker_b = MAF_B, ld_r2 = R2) %>%
  dplyr::mutate(ld_r2 = as.numeric(ld_r2)) %>%
  dplyr::filter(!is.na(ld_r2)) %>%
  dplyr::left_join(finemap,., by = "SNP") %>%
  dplyr::left_join(.,roi_genotype)

peak_roi_marker <- dplyr::filter(pr_roi_ld, POS == peakp)

ld_plot <- ggplot(pr_roi_ld, mapping = aes(x = POS/1000000, y = as.numeric(-log(P)), fill = ld_r2)) +
  theme_bw(15) +
  geom_point(shape = 23, size = 3) +
  geom_point(aes(y = -log(P)), shape = 23, size = 3, fill = "red",
             data = peak_roi_marker) +
  scale_fill_distiller(palette = "PuOr", name = bquote(r^2)) +
  labs(x = "Genomic Position (Mb)",
       y = expression(-log[10](italic(p))))

# ggsave(ld_plot, filename = finemap_plot.pdf", height = 4, width = 12)

print(ld_plot)

```


<br>
<br>

#### All variants in interval

```{r}

if(species == "c_elegans") {
  # All variants in interval
  var_info <- genes_in_region %>%
      dplyr::filter(START_POS == QTL_start & !is.na(WBGeneID)) %>%
      dplyr::select(MARKER, REF, ALT,WBGeneID,GENE_NAME, VARIANT_IMPACT, VARIANT_LOG10p, STRAIN_GENOTYPE, STRAIN) %>%
      dplyr::distinct() %>%
      dplyr::mutate(WBGeneID = paste0('<a href=https://wormbase.org/species/c_elegans/gene/', WBGeneID, '>', WBGeneID,'</a>' ),
                    STRAIN = stringr::str_replace_all(STRAIN, ",", ", "))
  
  DT::datatable(var_info, filter = "top", escape = FALSE, options = list(columnDefs = list(list(width = '20%', targets = c(8))))) %>%
      DT::formatStyle("STRAIN","white-space"="nowrap") %>%
      DT::formatStyle("VARIANT_IMPACT", backgroundColor = DT::styleEqual("HIGH", "yellow"))
} else {
  # All variants in interval
  var_info <- genes_in_region %>%
      dplyr::filter(START_POS == QTL_start & !is.na(WBGeneID)) %>%
      dplyr::select(MARKER, REF, ALT,WBGeneID,GENE_NAME, VARIANT_IMPACT, VARIANT_LOG10p, STRAIN_GENOTYPE, STRAIN) %>%
      dplyr::distinct() %>%
      dplyr::mutate(STRAIN = stringr::str_replace_all(STRAIN, ",", ", "))
  
  DT::datatable(var_info, filter = "top", escape = FALSE, options = list(columnDefs = list(list(width = '20%', targets = c(8))))) %>%
      DT::formatStyle("STRAIN","white-space"="nowrap") %>%
      DT::formatStyle("VARIANT_IMPACT", backgroundColor = DT::styleEqual("HIGH", "yellow"))
}


```

<br>
<br>

#### Mediation analysis

Mediation analysis was performed to analyze if gene expression variation is significantly correlated with the phenotype (overlap of phenotype QTL with expression QTL). Top candidates whose expression might mediate the phenotype QTL are shown below. (*Note: expression data currently unpublished*). For more information about mediation analysis, check out Evans and Andersen 2020 (PMID: 32385045).

```{r, eval = whether_mediate}

if(whether_mediate) {
    # plot mediation
    # plot
    summarized_med <- readr::read_tsv(glue::glue("../../{alg}/Mediation/{trait_name}_mediation_{stringr::str_to_lower(alg)}.tsv"))

    medprop_all <- summarized_med %>% 
      dplyr::mutate(e_peak=as.numeric(e_peak),
                    estimate=as.numeric(estimate),
                    prob=as.numeric(prob)) %>% 
      dplyr::filter(var == "prop_med") %>% 
      dplyr::filter(estimate >=0 & estimate<=1) %>% 
      dplyr::mutate(gene_gwas_qtl=paste(mediator,gwchr,gwpeak, sep = "_"))


    med  <- summarized_med %>% 
      dplyr::select(-c("var","estimate","ci_lower","ci_upper","prob")) %>% 
      dplyr::distinct() %>% 
      dplyr::mutate(e_peak=as.numeric(e_peak),
                    multi_abs_est=as.numeric(multi_abs_est),
                    multi_padjust=as.numeric(multi_padjust)) %>% 
      dplyr::mutate(gene_gwas_qtl=paste(mediator,gwchr,gwpeak, sep = "_")) %>% 
      dplyr::mutate(mediator=paste(mediator,mediator_gene,sep = "_")) %>% 
      dplyr::mutate(q99 = quantile(multi_abs_est, probs = 0.99)[[1]]) %>% 
      dplyr::mutate(prop_med=ifelse(gene_gwas_qtl %in% medprop_all$gene_gwas_qtl, "0 ≤ prop_med ≤ 1","NA"),
                    q99_mediator=ifelse(multi_abs_est>q99 & prop_med=="0 ≤ prop_med ≤ 1",mediator,NA)) %>% 
      dplyr::filter(multi_padjust<0.05)


    if (length(unique(med$q99_mediator)) > 1) {
      
      
    med_detailed_plot <- ggplot(med,aes(x=e_peak/1e6,y=multi_abs_est,color=q99_mediator )) +
      geom_point( size=2) +
      labs(x = "Genomic position (Mb)", y = "Mediation estimate") +
      scale_alpha_continuous(range = c(1, 0.1)) +
      geom_hline(aes(yintercept = q99), color = "grey") +
      theme_bw(10) +
      theme( panel.grid = element_blank(),
            panel.background = element_blank(), 
            #  legend.position = "none",
            axis.text = element_text(size=10,color="black"),
            axis.title = element_text(size=10, color="black"),
            strip.text = element_text(size=10, color = "black")) +
      facet_grid(eQTL_classification~gwchr,scales = "free_x")

    print(med_detailed_plot)

    # ggsave(med_detailed_plot, filename = paste(tr,"_emed_detailed_plot.png",sep = ""),  units = "in",height = 10, width =  10)

    } else {
      print("No significant mediators to report")
    }
} else {
    print("Mediation was not performed for this analysis.")
}

```

```{r, eval = !whether_mediate}

print(glue::glue("Mediation was not performed for this analysis"))

```

<br>
<br>

#### Divergent regions

We recently published about punctuated hyper-divergent regions in *C. elegans* (Lee *et al.* 2021 (PMID: 32385045)). Within these divergent regions, we are less confident about the variant calls and even the gene content between strains. For these reasons, if your QTL falls within a divergent region it may complicate your analyses and requires extra careful interpretation of fine-mapping results.

The following plot shows divergent regions for each strain across the QTL region. Strains are split by genotype at the peak marker. You should be careful if many strains are divergent, especially if most of the strains in the ALT group are divergent, for example.

```{r, out.width = '100%', fig.height=20, fig.width=20, eval = div}

# genome bin table
all_QTL_bins <- read.delim(glue::glue("../../{alg}/Divergent_and_haplotype/all_QTL_bins_{stringr::str_to_lower(alg)}.bed"),
                           header=FALSE, stringsAsFactors=FALSE) %>%
  dplyr::filter(V1 == QTL_chrom & V2 > QTL_start & V3 < QTL_end )

# if any of user supplied strains have divergent region with the QTL region
if (file.info(glue::glue("../../{alg}/Divergent_and_haplotype/all_QTL_div_{stringr::str_to_lower(alg)}.bed"))$size!=0) {
  all_QTL_div <- read.delim(glue::glue("../../{alg}/Divergent_and_haplotype/all_QTL_div_{stringr::str_to_lower(alg)}.bed"),
                            header=FALSE, stringsAsFactors=FALSE) %>%
    dplyr::filter(V4 %in% unique(processed_mapping$strain)) %>%
    dplyr::filter(V1 == QTL_chrom & V2 > QTL_start & V3 < QTL_end ) %>%
    dplyr::mutate(length = (V3 - V2)/1000) %>%
    tibble::rownames_to_column(var = "region")
  
} else {
  # create an empty dataframe
  all_QTL_div <- data.frame(V1=character(),
                            V2=integer(), 
                            V3=integer(), 
                            V4=character(), 
                            V5=integer(), 
                            length=integer(),
                            region=character(),
                            stringsAsFactors=FALSE)
}

# combine data

newtest <- NULL
for(i in 1:nrow(all_QTL_div)) {
  df <- all_QTL_div %>%
    dplyr::slice(rep(i, each = all_QTL_div$length[i])) %>%
    tibble::rownames_to_column() %>%
    dplyr::mutate(start = V2+(as.numeric(rowname)-1)*1000,
                  end = start + 1000) %>%
    dplyr::select(CHROM = V1, start,end, strain = V4) %>%
    dplyr::mutate(div = 1)
  
  newtest <- newtest %>%
    dplyr::bind_rows(df)
}

strains <- data.frame(strain = unique(processed_mapping$strain)[!is.na(unique(processed_mapping$strain))]) %>%
  dplyr::mutate(CHROM = QTL_chrom)

strain_bins <- all_QTL_bins %>%
  dplyr::select(CHROM = V1, start = V2, end = V3) %>%
  dplyr::full_join(strains) %>%
  dplyr::left_join(newtest) %>%
  dplyr::mutate(div = ifelse(is.na(div), 0, div))

# add ref/alt
ref_alt <- processed_mapping %>%
  dplyr::filter(peakPOS == QTL_Peak) %>% 
  dplyr::select(strain, allele) %>%
  dplyr::mutate(geno_group = ifelse(allele==1, "ALT", "REF"))  %>% 
  dplyr::group_by(geno_group) %>%
  dplyr::mutate(geno_group_count = paste0(geno_group, " (", dplyr::n(), ")")) %>%
  dplyr::select(-allele)

ref_alt_div <- strain_bins %>%
  dplyr::left_join(ref_alt)

# decide whether there is enough strains with divergent region to split phenotype genotype plot
if (nrow(dplyr::filter(ref_alt_div, div == 1)) > nrow(ref_alt_div)/20 ) {
    split_by_div = TRUE
} else { split_by_div = FALSE }


# plot
for.div.plot <- ref_alt_div %>%
  dplyr::left_join(ref_alt) %>%
  dplyr::mutate(div = case_when(
    div == 0 ~ "Not Divergent",
    div == -1 ~ "No Divergent Information",
    div == 1 ~ "Divergent")) %>%
  dplyr::mutate(color = case_when(
    div == "Not Divergent" ~ "grey",
    div == "No Divergent Information" ~ "darkblue",
    div == "Divergent" ~ "darkred"))
div.palette <- for.div.plot$color
names(div.palette) <- for.div.plot$div

ggplot2::ggplot(for.div.plot) +
  aes(x = start/1e6, xend = end/1e6, yend = strain, y = strain) +
  ggplot2::geom_segment(aes(color = div), size = 2) +
  theme_bw(12) +
  ggplot2::theme(panel.spacing = unit(1, "lines"),
          strip.text = element_text(size = 30, face = "bold"),
          axis.text.x = element_text(size = 30),
          axis.text.y = element_text(size = 10),
          axis.title = element_text(size = 30, face = "bold"),
          axis.ticks.y = element_blank(),
          panel.grid = element_blank(),
          legend.text = element_text(size = 20)) +
  scale_color_manual(values = div.palette, name = "Divergence") +
  facet_grid(geno_group_count~., scales = "free_y", space = "free_y") +
  geom_vline(xintercept = QTL_Peak/1e6, linetype = "dashed", size = 2) +
  labs(x = "Genomic position (Mb)", y = "Strain")

```

```{r, eval = !div}

print(glue::glue("Divergent regions were not calculated."))
split_by_div = FALSE

```

```{r, eval=split_by_div, results="asis"}
cat("\n#### Phenotype x Genotype split by divergent region")
cat("  \n")

cat("\n The following plot shows the phenotype-by-genotype splits for strains that are divergent from N2 or not divergent from N2. Again, you should be careful interpreting QTL that might be driven by divergent region structure.")

```

```{r, eval=split_by_div}

nested.pxg.dat <- processed_mapping %>%
    dplyr::filter(algorithm == alg) %>%
    dplyr::filter(!is.na(peak_id)) %>%
    dplyr::select(CHROM, marker, trait, startPOS, peakPOS, endPOS, AF1, value, strain, allele, peak_id) %>%
    dplyr::distinct() %>%
    dplyr::filter(CHROM==QTL_chrom, peakPOS==QTL_Peak) %>%  
    dplyr::mutate(value=as.numeric(value)) %>% 
    dplyr::inner_join(ref_alt_div, by="strain") %>%
    dplyr::distinct(marker, strain, allele, div, .keep_all = T) %>%
    dplyr::group_by(marker, strain, allele) %>%
    dplyr::filter(div == max(div)) %>%
    dplyr::ungroup() %>%
    dplyr::mutate(startPOS = startPOS/1000000,
                  peakPOS = peakPOS/1000000,
                  endPOS = endPOS/1000000) %>%
    # dplyr::left_join(.,sweep.chrom.pivot) %>% # don't have this file yet
    # dplyr::group_by(trait, peak_id) %>%
    # dplyr::recode(allele, "-1" = "REF", "1" = "ALT") %>%
    dplyr::mutate(allele = dplyr::case_when(allele == "-1" ~ "REF",
                                            allele == "1" ~ "ALT",
                                            TRUE ~ "NA"),
                  allele = factor(allele, levels = c("REF", "ALT")))

plot <- nested.pxg.dat %>%
    dplyr::filter(allele != "NA" | !is.na(allele)) %>%
    droplevels() %>%
    ggplot2::ggplot(mapping = aes(x = allele, y = value)) +
    ggplot2::theme_bw(base_size = 12) +
    ggplot2::geom_violin(aes(fill = allele), alpha = 0.5, scale = "count", draw_quantiles = c(0.25, 0.5, 0.75)) +
    ggplot2::scale_fill_manual(values = c("REF" = "#A79F92", "ALT" = "mediumpurple4"), guide = FALSE) +
    ggbeeswarm::geom_beeswarm(size = 1.1) + 
    ggrepel::geom_text_repel(data = nested.pxg.dat[which(nested.pxg.dat$strain %in% c("N2","CB4856")),], aes(label = strain), colour = "black", 
                  box.padding = 1, point.padding = 0.1) + 
    ggplot2::theme(legend.position = "bottom") +
    ggplot2::facet_wrap(~div, labeller = labeller(div = 
                                                c("-1" = "no div data",
                                                  "0" = "not divergent",
                                                  "1" = "divergent from N2")) ) +
    ggplot2::labs(y = "Trait Value", 
                  x = "Genotype")


print(plot)

```

<br>
<br>

#### Haplotype

The following plot shows the genome-wide haplotype (genetic relatedness) of mapped strains split by REF or ALT genotype. This plot can be useful to help identify how many unique haplotypes are present in the REF or ALT groups. If you want to choose parent strains for a NIL cross to validate this QTL, you might want to choose strains in the major haplotype of the REF/ALT groups that also have distinct phenotypes.

```{r, fig.height=20, fig.width=20, eval = div}
# if (file.info("all_QTL_div.bed")$size!=0) {

haplotype_in_QTL_region <- read.delim(glue::glue("../../{alg}/Divergent_and_haplotype/haplotype_in_QTL_region_{stringr::str_to_lower(alg)}.txt"), 
                                      header=FALSE, stringsAsFactors=FALSE) %>% 
    dplyr::select(-V12)

names(haplotype_in_QTL_region) <- c("chromosome", "start", "stop", "haplotype", "isotype", "plotpoint", "segment", "color", "Chrom", "Region_start", "Region_end")


########### filter only tested strains

########### keep rows for this QTL region
############ keep relavent columns for plotting

plot_df <- haplotype_in_QTL_region %>%
    dplyr::filter(Chrom == QTL_chrom & Region_start == QTL_start) %>% 
    dplyr::select(chromosome, start, stop, haplotype, isotype, plotpoint, segment, color)


########### sort by haplotype of 1st and last segment

plot_df_seg_1 <- plot_df %>% 
    dplyr::arrange(isotype, start) %>%
    dplyr::distinct(isotype, .keep_all = TRUE) %>% 
    dplyr::select(isotype, haplotype) %>% 
    dplyr::rename(haplotype_start = haplotype)

plot_df_seg_last <- plot_df %>% 
    dplyr::arrange(isotype, -start) %>%
    dplyr::distinct(isotype, .keep_all = TRUE) %>% 
    dplyr::select(isotype, haplotype) %>% 
    dplyr::rename(haplotype_end = haplotype) 

plot_df_sort_order <- plot_df_seg_1 %>%
    dplyr::inner_join(plot_df_seg_last) %>%
    dplyr::arrange(haplotype_start, haplotype_end) %>% 
    dplyr::mutate(plotpoint2=row_number()) %>% 
    dplyr::inner_join(ref_alt, by=c("isotype"="strain")) %>% 
    dplyr::group_by(geno_group) %>% 
    dplyr::mutate(plotpoint3 = rank(plotpoint2)) %>% 
    dplyr::ungroup() %>% 
    dplyr::select(isotype, plotpoint3, geno_group)


########## add new plotpoint back to plot_df 

plot_df <- plot_df %>%
    dplyr::inner_join(plot_df_sort_order, by="isotype")

########## relevel REF/ALT

plot_df$geno_group <- factor(plot_df$geno_group, levels=c("REF","ALT"))


#=======================#
# Normal haplotype plot #
#=======================#


mcolor_grp <- plot_df %>% 
    dplyr::select(haplotype, color) %>% 
    dplyr::distinct()
mcolor <- mcolor_grp$color

# mcolor_grp2 <- plot_df %>% 
#     dplyr::select(haplotype, color_new) %>% 
#     dplyr::distinct()
# mcolor2 <- mcolor_grp2$color_new

names(mcolor) <- mcolor_grp$haplotype

strain_labels <- plot_df %>%
    dplyr::select(isotype, plotpoint3) %>% 
    dplyr::distinct()

plot_df <- plot_df %>%
    dplyr::arrange((plotpoint3))

plot_df$isotype <- factor(plot_df$isotype, unique(plot_df$isotype))

p1 <- ggplot2::ggplot(filter(plot_df),
            ggplot2::aes(x = start/1e6, xend = stop/1e6,
                 y = isotype, yend = isotype,
                 color = haplotype)) +
    ggplot2::geom_segment(size = 3) +
    ggplot2::scale_color_manual(values = mcolor) +
    ggplot2::theme_bw() +
    ggplot2::coord_cartesian(xlim=c(QTL_start/1e6, QTL_end/1e6)) +
    ggplot2::theme(legend.position="none") +
    ggplot2::geom_vline(xintercept = QTL_Peak/1e6, size = 2) +
    ggplot2::facet_grid(geno_group ~ ., scales = "free", space = "free") + 
    ggplot2::theme(panel.spacing = unit(1, "lines"),
                   strip.text = element_text(size = 30, face = "bold"),
                   axis.text.x = element_text(size = 30),
                   axis.text.y = element_text(size = 10),
                   axis.title = element_text(size = 30, face = "bold")) +
    ggplot2::labs(x = "Genomic position (Mb)", y = "Isotype")


#ggsave(paste0("haplotype_", QTL_Chrom, ".", QTL_Region_start, ".", QTL_Region_end, ".png"), p, height = 30, width = 8)


# p2 <- ggplot2::ggplot(filter(plot_df),
#              ggplot2::aes(x = start, xend = stop,
#                  y = isotype, yend = isotype,
#                  color = haplotype)) +
#     ggplot2::geom_segment(size = 3) +
#     ggplot2::scale_color_manual(values = mcolor2) +
#     ggplot2::xlab("Position") +
#     ggplot2::theme_bw() +
#     ggplot2::coord_cartesian(xlim=c(QTL_start, QTL_end)) +
#     ggplot2::theme(legend.position="none") +
#     ggplot2::geom_vline(xintercept = QTL_Peak) +
#     ggplot2::facet_grid(geno_group ~ ., scales = "free", space = "free") + 
#     ggplot2::theme(panel.spacing = unit(1, "lines"))

print(p1)
# cowplot::plot_grid(p1, p2, ncol=2)

# ggplotly(p1) # cannot get the grid panel size to maintain the scale. to be fixed.
```

```{r, eval = !div}

print(glue::glue("Haplotypes not calculated."))

```

<br>
