
### `r region_name`

#### Fine mapping

Fine mapping was performed by evaluating the genotype-phenotype relationship for variants nearby the QTL identified from GWA mapping using a vcf containing imputed variants to avoid removing variants with missing genotype information for one or a few strains. Only SNVs were considered in this mapping.

```{r}
# read in genes in region and pval
genes_in_region <- data.table::fread(glue::glue("Fine_Mappings/Data/{trait_name}_{QTL_chrom}_{QTL_start}-{QTL_end}_bcsq_genes.tsv"))
    
gene_df <- genes_in_region %>%
    dplyr::filter(START_POS == unique(genes_in_region$START_POS)) %>%
    dplyr::distinct(WBGeneID, PEAK_MARKER, CHROM, STRAND, TRANSCRIPTION_START_POS, 
                    TRANSCRIPTION_END_POS, START_POS, END_POS, VARIANT_LOG10p) %>%
    dplyr::group_by(WBGeneID) %>%
    dplyr::mutate(VARIANT_LOG10p = max(VARIANT_LOG10p, na.rm = T)) %>%
    dplyr::distinct()

peak_variant <- as.numeric(strsplit(unique(gene_df$PEAK_MARKER), split = ":")[[1]][2])

variant_df <- genes_in_region %>%
    dplyr::filter(START_POS == unique(genes_in_region$START_POS)) %>%
    dplyr::distinct(CHROM, POS, VARIANT_LOG10p, VARIANT_IMPACT)

variant_df$VARIANT_IMPACT[is.na(variant_df$VARIANT_IMPACT)] <- "Intergenic"

xs <- unique(gene_df$START_POS)
xe <- unique(gene_df$END_POS)
cq <- unique(gene_df$CHROM)

max_logp <- unique(max(variant_df$VARIANT_LOG10p, na.rm = T))/150

gene_plot <- ggplot(gene_df) +
    geom_vline(aes(xintercept = peak_variant/1e6),
               linetype=3, color = "cyan")+
    geom_segment(aes(x = ifelse(STRAND == "+", TRANSCRIPTION_START_POS/1e6, TRANSCRIPTION_END_POS/1e6),
                     xend = ifelse(STRAND == "+", TRANSCRIPTION_END_POS/1e6, TRANSCRIPTION_START_POS/1e6),
                     y = VARIANT_LOG10p,
                     yend = VARIANT_LOG10p),
                 arrow = arrow(length = unit(5, "points")), size = 1) +
    geom_segment(aes(x = POS/1e6,
                     xend = POS/1e6,
                     y = VARIANT_LOG10p+max_logp,
                     yend = VARIANT_LOG10p-max_logp,
                     color = VARIANT_IMPACT), data = variant_df) +
    scale_color_manual(values = c("MODIFIER" = "gray50",
                                  "LOW" = "gray30",
                                  "MODERATE" = "orange",
                                  "HIGH" = "red",
                                  "Intergenic" = "gray80"),
                       breaks = c("HIGH", "MODERATE", "LOW", "MODIFIER", "Intergenic"),
                       name = "EFFECT")+
    labs(x = "Genomic Position (Mb)",
         y = expression(-log[10](italic(p))))+
    theme_bw(18)+
    xlim(c(xs/1e6, xe/1e6)) +
    theme(legend.position = "top",
          panel.grid = element_blank())

print(gene_plot)
    
    # ggsave(gene_plot,
    #        filename = glue::glue("{analysis_trait}_{cq}_{xs}-{xe}_gene_plot_{ann_type}.pdf"),
    #        height=10, width = 14)



```

```{r}

finemap <- data.table::fread(glue::glue("Fine_Mappings/Data/{trait_name}.{QTL_chrom}.{QTL_start}.{QTL_end}.finemap_inbred.fastGWA")) %>%
  dplyr::mutate(CHR = as.factor(CHR),
                SNP = as.factor(SNP),
                POS = as.numeric(POS),
                AF1 = as.numeric(AF1),
                P = as.numeric(P))
ROI.geno.matrix <- data.table::fread(glue::glue("Fine_Mappings/Data/{trait_name}.{QTL_chrom}:{QTL_start}-{QTL_end}.ROI_Genotype_Matrix.tsv"))
ROI.LD <- data.table::fread(glue::glue("Fine_Mappings/Data/{trait_name}.{QTL_chrom}.{QTL_start}.{QTL_end}.LD.tsv"))

peakp <- unique(ROI.LD$BP_A)

finemap_peaks <- na.omit(finemap) %>%
  dplyr::filter(POS == peakp)

finemap_peaks$POS <- as.numeric(finemap_peaks$POS)

# add strain genotype to ld file for snpeff later
roi_genotype <- ROI.geno.matrix %>%
  tidyr::gather(strain, allele, -c(CHROM:ALT)) %>%
  dplyr::mutate(allele = ifelse(allele == -1, "REF", ifelse(allele == 1, "ALT", NA))) %>%
  dplyr::group_by(CHROM, POS, REF, ALT, allele) %>%
  dplyr::summarize(strains = paste(strain, collapse = ",")) 

pr_roi_ld <- ROI.LD %>%
  dplyr::mutate(peak_marker = gsub("_", ":", unique(finemap_peaks$SNP))) %>%
  dplyr::mutate(SNP = SNP_B) %>%
  dplyr::select(peak_marker, peak_maf = MAF_A, SNP, maf_marker_b = MAF_B, ld_r2 = R2) %>%
  dplyr::mutate(ld_r2 = as.numeric(ld_r2)) %>%
  dplyr::filter(!is.na(ld_r2)) %>%
  dplyr::left_join(finemap,., by = "SNP") %>%
  dplyr::left_join(.,roi_genotype)

peak_roi_marker <- dplyr::filter(pr_roi_ld, POS == peakp)

ld_plot <- ggplot(pr_roi_ld, mapping = aes(x = POS/1000000, y = as.numeric(-log(P)), fill = ld_r2)) +
  theme_bw(15) +
  geom_point(shape = 23, size = 3) +
  geom_point(aes(y = -log(P)), shape = 23, size = 3, fill = "red",
             data = peak_roi_marker) +
  scale_fill_distiller(palette = "PuOr", name = bquote(r^2)) +
  labs(x = "Genomic Position (Mb)",
       y = expression(-log[10](italic(p))))

# ggsave(ld_plot, filename = finemap_plot.pdf", height = 4, width = 12)

print(ld_plot)

```

#### Divergent regions

```{r}

# check whether there are user tested strains that have no divergent region data
div_strain_list <- read.delim("Divergent_and_haplotype/div_isotype_list.txt", header=FALSE, stringsAsFactors=FALSE) 

# where does this come from?
user_strain_pheno_list <- processed_mapping %>%
    dplyr::select(strain) %>%
    dplyr::distinct() %>%
    na.omit()

w_div_data_strain <- user_strain_pheno_list %>%
    dplyr::inner_join(div_strain_list, by=c("strain"="V1"))
no_data_strain <- user_strain_pheno_list %>%
    dplyr::anti_join(div_strain_list, by=c("strain"="V1"))

print("User-tested strains that do not have divergent region data (shown in blue below): ")
print(no_data_strain$strain)

# genome bin table
all_QTL_bins <- read.delim("Divergent_and_haplotype/all_QTL_bins.bed", header=FALSE, stringsAsFactors=FALSE)

# take genomic 1kb bins within the QTL region
# format to the same as div regions
# during spread, this "all_bins" will dictate number of rows in resulting dataframe
region_QTL_bins <- all_QTL_bins %>%
    dplyr::filter(V1 == QTL_chrom & V2 > QTL_start & V3 < QTL_end ) %>% 
    dplyr::mutate(V4 = "all_bins", V5 = 0)


# if any of user supplied strains have divergent region with the QTL region
if (file.info("Divergent_and_haplotype/all_QTL_div.bed")$size!=0) {
    
    all_QTL_div <- read.delim("Divergent_and_haplotype/all_QTL_div.bed", header=FALSE, stringsAsFactors=FALSE)
    
} else {
    # create an empty dataframe
    all_QTL_div <- data.frame(V1=character(),
                              V2=integer(), 
                              V3=integer(), 
                              V4=character(), 
                              V5=integer(), 
                              stringsAsFactors=FALSE)
}


strain_QTL_div <- all_QTL_div %>%
    dplyr::filter(V4 %in% user_strain_pheno_list$strain)


# take divergent bins within the QTL region, for all user tested strains
# append all_bins within QTL region to the QTL_div 
# spread to short form so each strain is a column, each bin in QTL region is a row, all bins are included.
region_QTL_div <- strain_QTL_div %>%
    dplyr::filter(V1==QTL_chrom & V2 > QTL_start & V3 < QTL_end ) %>%
    dplyr::bind_rows(region_QTL_bins)  %>% 
    tidyr::spread(V4, V5, fill=0) %>% 
    dplyr::select(-all_bins)

# add back user-tested strains that have divergent data but are not divergent within the QTL peak region, therefore left out in previous step
w_div_data_strain_missed <- w_div_data_strain$strain[!(w_div_data_strain$strain %in% names(region_QTL_div))]


for (x in w_div_data_strain_missed){
    region_QTL_div[[x]] <- 0
}


# add back user-tested strains that have no divergent data
for (x in no_data_strain$strain){
    region_QTL_div[[x]] <- -1
}


# check that all user-tested strains are include now
if (ncol(region_QTL_div) - 3 != nrow(user_strain_pheno_list)){ print("strain number don't match!!!") }


m <- as.matrix(dplyr::select(region_QTL_div,-V1,-V2,-V3))

# transpose to the format that heatmap needs
m2 <- t(m)
colnames(m2) <- region_QTL_div$V2


# add REF/ALT annotation
ref_alt <- processed_mapping %>%
    dplyr::filter(peakPOS == QTL_Peak) %>% 
    dplyr::select(strain, allele) %>% 
    dplyr::filter(strain %in% row.names(m2))

alt_count <- data.frame(table(ref_alt$allele)) %>% 
    dplyr::filter(Var1==1)
ref_count <- data.frame(table(ref_alt$allele)) %>% 
    dplyr::filter(Var1==-1)

ref_alt <- ref_alt %>%
    dplyr::mutate(geno_group = ifelse(allele==1, "ALT", "REF"))  %>% 
    dplyr::mutate(geno_group_count = ifelse(allele==1, paste0("ALT(", alt_count[1,2], ")"), paste0("REF(", ref_count[1,2], ")"))) %>% 
    dplyr::select(-allele)

# make sure REF/ALT annotation is in the same order as row.names(m2)
# that's how complexHeatmap combine them
ref_alt_annotation <- data.frame(strain=row.names(m2)) %>% 
    dplyr::left_join(ref_alt, by=c("strain" = "strain"))


# add div and write out table
# this corresponds to the sort order of rows for heatmap
peak_bin <- floor(QTL_Peak/1000)*1000
ref_alt_annotation <- ref_alt_annotation %>%
    dplyr::rowwise() %>%
    mutate(is_div = m2[strain, as.character(peak_bin)])


# decide whether there is enough strains with divergent region to split phenotype genotype plot
if (nrow(dplyr::filter(ref_alt_annotation, is_div == 1)) > nrow(ref_alt_annotation)/20 ) {
    split_by_div = TRUE
} else { split_by_div = FALSE }

######################### make the plot
# set color for heatmap
colors <- structure(c("indianred1","grey90","dodgerblue3"), names = c("1", "0", "-1")) 

# keep only existing groups for legend
legend_mapping <- data.frame(div_group=c(0, 1, -1), div_label=c("not divergent", "divergent from N2", "no divergent data" )) %>% 
    dplyr::filter(div_group %in% unique(c(m2)))

# create x-axis label
x_axis_start <- colnames(m2)[1]
x_axis_peak <- round(QTL_Peak, digits = -3)
x_axis_end <- colnames(m2)[ncol(m2)]

genomic_coor_text <- as.numeric(colnames(m2))
genomic_coor_text[!genomic_coor_text %in% c(x_axis_start,x_axis_peak, x_axis_end)] <- ""

test <- data.frame(m2) %>%
    tibble::rownames_to_column(var = "strain") %>%
    tidyr::gather(pos, div, -strain) %>%
    dplyr::mutate(pos = as.numeric(gsub("X", "", pos)),
                  div = as.character(div)) %>%
    dplyr::full_join(ref_alt_annotation) %>%
    dplyr::arrange(geno_group, desc(div))
test$strain <- factor(test$strain, levels = unique(test$strain))
test$geno_group_count <- factor(test$geno_group_count, levels = rev(unique(test$geno_group_count)))

test %>%
    ggplot2::ggplot(.) +
    aes(x = pos, y = strain, fill = div) +
    geom_tile() +
    theme_bw() +
    theme(axis.text.y = element_blank(),
          axis.ticks.y = element_blank(),
          panel.grid = element_blank()) +
    scale_fill_manual(values = c("1" = "indianred1","0" = "grey90","-1" = "dodgerblue3"), 
                      labels = c("not divergent", "divergent from N2", "no divergent data")) +
    facet_grid(geno_group_count~., scales = "free_y", space = "free_y") +
    geom_vline(xintercept = round(QTL_Peak, digits = -3), linetype = "dashed") +
    labs(x = "Genomic position (Mb)", y = "Strain")


# x_axis_label <- ComplexHeatmap::HeatmapAnnotation(
#     year = ComplexHeatmap::anno_text(genomic_coor_text, rot = 0, location = unit(1, "npc"), just = "top"))
# 
# # add empty space on top of plot so the text "peak" will be within plottin area and show
# peak_label_space <- ComplexHeatmap::HeatmapAnnotation(foo = ComplexHeatmap::anno_empty(border = FALSE, height = unit(1, "cm")))
# 
# 
# ht <- ComplexHeatmap::Heatmap(m2, name="div_region",
#               row_split=factor(ref_alt_annotation$geno_group_count,
#                                levels=c(paste0("REF(", ref_count[1,2], ")"), paste0("ALT(", alt_count[1,2], ")"))), row_gap = unit(3, "mm"),
#               cluster_columns = FALSE, 
#               show_row_dend = FALSE, 
#               cluster_row_slices=FALSE, 
#               col = colors, 
#               heatmap_legend_param = list( title="",
#                                            at = legend_mapping$div_group,
#                                            labels = as.character(legend_mapping$div_label)),
#               show_column_names = FALSE,
#               show_row_names = FALSE,
#               bottom_annotation = x_axis_label,
#               top_annotation = peak_label_space)
# 
# ComplexHeatmap:: draw(ht)
# 
# ComplexHeatmap::decorate_heatmap_body("div_region", {
#     i = which(colnames(m2) == round(QTL_Peak, digits = -3))
#     x = i/ncol(m2)
#     grid.lines(c(x, x), c(0, 1), gp = gpar(lwd = 2, lty = 2))
#     grid.text("QTL peak", x, unit(1, "npc") + unit(5, "mm"))
# }, row_slice=1)
# 
# ComplexHeatmap::decorate_heatmap_body("div_region", {
#     i = which(colnames(m2) == round(QTL_Peak, digits = -3))
#     x = i/ncol(m2)
#     grid.lines(c(x, x), c(0, 1), gp = gpar(lwd = 2, lty = 2))
# }, row_slice=2)

#dev.copy(pdf,'divergent_regions.pdf')
#dev.off()

```

<br>
<br>

#### Haplotype

<br>
```{r, fig.height=20, fig.width=20}
# if (file.info("all_QTL_div.bed")$size!=0) {

haplotype_in_QTL_region <- read.delim("Divergent_and_haplotype/haplotype_in_QTL_region.txt", header=FALSE, stringsAsFactors=FALSE) %>% 
    dplyr::select(-V13)

names(haplotype_in_QTL_region) <- c("chromosome", "start", "stop", "haplotype", "isotype", "plotpoint", "segment", "color", "Chrom", "Region_start", "Region_end")


########### filter only tested strains

########### keep rows for this QTL region
############ keep relavent columns for plotting

plot_df <- haplotype_in_QTL_region %>%
    dplyr::filter(Chrom == QTL_chrom & Region_start == QTL_start) %>% 
    dplyr::select(chromosome, start, stop, haplotype, isotype, plotpoint, segment, color)


########### sort by haplotype of 1st and last segment

plot_df_seg_1 <- plot_df %>% 
    dplyr::arrange(isotype, start) %>%
    dplyr::distinct(isotype, .keep_all = TRUE) %>% 
    dplyr::select(isotype, haplotype) %>% 
    dplyr::rename(haplotype_start = haplotype)

plot_df_seg_last <- plot_df %>% 
    dplyr::arrange(isotype, -start) %>%
    dplyr::distinct(isotype, .keep_all = TRUE) %>% 
    dplyr::select(isotype, haplotype) %>% 
    dplyr::rename(haplotype_end = haplotype) 

plot_df_sort_order <- plot_df_seg_1 %>%
    dplyr::inner_join(plot_df_seg_last) %>%
    dplyr::arrange(haplotype_start, haplotype_end) %>% 
    dplyr::mutate(plotpoint2=row_number()) %>% 
    dplyr::inner_join(ref_alt, by=c("isotype"="strain")) %>% 
    dplyr::group_by(geno_group) %>% 
    dplyr::mutate(plotpoint3 = rank(plotpoint2)) %>% 
    dplyr::ungroup() %>% 
    dplyr::select(isotype, plotpoint3, geno_group)


########## add new plotpoint back to plot_df 

plot_df <- plot_df %>%
    dplyr::inner_join(plot_df_sort_order, by="isotype")

########## relevel REF/ALT

plot_df$geno_group <- factor(plot_df$geno_group, levels=c("REF","ALT"))


#=======================#
# Normal haplotype plot #
#=======================#


mcolor_grp <- plot_df %>% 
    dplyr::select(haplotype, color) %>% 
    dplyr::distinct()
mcolor <- mcolor_grp$color

# mcolor_grp2 <- plot_df %>% 
#     dplyr::select(haplotype, color_new) %>% 
#     dplyr::distinct()
# mcolor2 <- mcolor_grp2$color_new

names(mcolor) <- mcolor_grp$haplotype

strain_labels <- plot_df %>%
    dplyr::select(isotype, plotpoint3) %>% 
    dplyr::distinct()

plot_df <- plot_df %>%
    dplyr::arrange((plotpoint3))

plot_df$isotype <- factor(plot_df$isotype, unique(plot_df$isotype))

p1 <- ggplot2::ggplot(filter(plot_df),
            ggplot2::aes(x = start, xend = stop,
                 y = isotype, yend = isotype,
                 color = haplotype)) +
    ggplot2::geom_segment(size = 3) +
    ggplot2::scale_color_manual(values = mcolor) +
    ggplot2::xlab("Position") +
    ggplot2::theme_bw() +
    ggplot2::coord_cartesian(xlim=c(QTL_start, QTL_end)) +
    ggplot2::theme(legend.position="none") +
    ggplot2::geom_vline(xintercept = QTL_Peak) +
    ggplot2::facet_grid(geno_group ~ ., scales = "free", space = "free") + 
    ggplot2::theme(panel.spacing = unit(1, "lines"))


#ggsave(paste0("haplotype_", QTL_Chrom, ".", QTL_Region_start, ".", QTL_Region_end, ".png"), p, height = 30, width = 8)


# p2 <- ggplot2::ggplot(filter(plot_df),
#              ggplot2::aes(x = start, xend = stop,
#                  y = isotype, yend = isotype,
#                  color = haplotype)) +
#     ggplot2::geom_segment(size = 3) +
#     ggplot2::scale_color_manual(values = mcolor2) +
#     ggplot2::xlab("Position") +
#     ggplot2::theme_bw() +
#     ggplot2::coord_cartesian(xlim=c(QTL_start, QTL_end)) +
#     ggplot2::theme(legend.position="none") +
#     ggplot2::geom_vline(xintercept = QTL_Peak) +
#     ggplot2::facet_grid(geno_group ~ ., scales = "free", space = "free") + 
#     ggplot2::theme(panel.spacing = unit(1, "lines"))

print(p1)
# cowplot::plot_grid(p1, p2, ncol=2)

# ggplotly(p1) # cannot get the grid panel size to maintain the scale. to be fixed.
```
<br>
<br>

```{r, eval=split_by_div, results="asis"}
cat("\n#### Phenotype x Genotype split by divergent region")
cat("  \n")
```

```{r, eval=split_by_div}

nested.pxg.dat <- processed_mapping %>%
    dplyr::filter(!is.na(peak_id)) %>%
    dplyr::select(CHROM, marker, trait, startPOS, peakPOS, endPOS, AF1, value, strain, allele, peak_id) %>%
    dplyr::distinct() %>%
    dplyr::filter(CHROM==QTL_chrom, peakPOS==QTL_Peak) %>%  
    dplyr::mutate(value=as.numeric(value)) %>% 
    dplyr::inner_join(ref_alt_annotation, by="strain") %>%
    dplyr::mutate(startPOS = startPOS/1000000,
                  peakPOS = peakPOS/1000000,
                  endPOS = endPOS/1000000) %>%
    # dplyr::left_join(.,sweep.chrom.pivot) %>% # don't have this file yet
    # dplyr::group_by(trait, peak_id) %>%
    # dplyr::recode(allele, "-1" = "REF", "1" = "ALT") %>%
    dplyr::mutate(allele = dplyr::case_when(allele == "-1" ~ "REF",
                                            allele == "1" ~ "ALT",
                                            TRUE ~ "NA"),
                  allele = factor(allele, levels = c("REF", "ALT")))
    
CB.N2.code <- nested.pxg.dat %>%
    dplyr::select(strain, allele) %>%
    dplyr::distinct() %>%
    dplyr::filter(strain %in% c("CB4856","N2","PD1074")) %>%
    droplevels()
    
if(length(levels(CB.N2.code$allele)) < 2){
  pal <- c("#726E75","#720E07")
} else {
  pal <- c("#FFA500","#0000ff")
}


plot <- nested.pxg.dat %>%
    dplyr::filter(allele != "NA" | !is.na(allele)) %>%
    droplevels() %>%
    ggplot2::ggplot(mapping = aes(x = allele, y = value)) +
    ggplot2::theme_bw(base_size = 12) +
    ggplot2::geom_violin(aes(fill = allele), alpha = 0.8, scale = "count", draw_quantiles = c(0.25, 0.5, 0.75)) +
    ggplot2::scale_fill_manual(values = pal, guide = FALSE) +
    ggbeeswarm::geom_beeswarm(size = 1.1) + 
    ggrepel::geom_text_repel(data = nested.pxg.dat[which(nested.pxg.dat$strain %in% c("N2","CB4856")),], aes(label = strain), colour = "black", 
                  box.padding = 1, point.padding = 0.1) + 
    ggplot2::theme(legend.position = "bottom") +
    ggplot2::facet_wrap(~is_div, labeller = labeller(is_div = 
                                                c("-1" = "no div data",
                                                  "0" = "not divergent",
                                                  "1" = "divergent from N2")) ) +
    ggplot2::labs(y = "Trait Value", 
                  x = "Genotype")

print(plot)

```


