---
title: "NemaScan Report for TRAIT_NAME_HOLDER"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_depth: 5
    toc_float:
      collapsed: true
      smooth_scroll: true
---

<br>

## Usage

* This R Markdown file imports data tables within Analysis_Results-Date folder and will knit without any modification.
<br>
* ggsave code is kept for users to make customized plots. 
<br>
<br>
<br>


```{r global_options, include=FALSE}
knitr::opts_chunk$set(echo=FALSE, warning=FALSE, message=FALSE)
options(knitr.duplicate.label = 'allow')
knitr::opts_knit$set(progress = FALSE)
```



```{r include=FALSE}
library.path <- .libPaths()
library(dplyr, lib.loc = library.path)
library(tidyr, lib.loc = library.path)
library(ggplot2, lib.loc = library.path)
library(stringr, lib.loc = library.path)
library(readr, lib.loc = library.path)
library(glue, lib.loc = library.path)
library(purrr, lib.loc = library.path)
library("plotly", lib.loc = library.path)
library("DT", lib.loc = library.path)
library("ggbeeswarm", lib.loc = library.path)
library("knitr", lib.loc = library.path)
library("ggrepel", lib.loc = library.path)
library("genetics", lib.loc = library.path)
library("ggnewscale", lib.loc = library.path)
library("cowplot", lib.loc = library.path)
library("stringr", lib.loc = library.path)

```

## Strain Issues

All strain names were converted to the corresponding isotype name, which can be looked up here: https://elegansvariation.org/strains/isotype_list. If you submitted replicate data, replicates for a given isotype were averaged to one mean value.

```{r}

strain_issues <- readLines("../../Phenotypes/strain_issues.txt")

new <- grep("WARNING:|Note:", strain_issues, value = T)
if(length(new) == 0) {
  print("No strain issues to report")
} else {
  print(new)
}

```


```{r}
# load trait name
# in nextflow use sed to edit this field and make a copy of the .rmd for each trait
trait_name <- "TRAIT_NAME_HOLDER"

species <- "SPECIES"

# load independent tests result
total_independent_tests <- read.table("../../Genotype_Matrix/total_independent_tests.txt", quote="\"", comment.char="", stringsAsFactors=FALSE)

independent_test_cutoff <- -log10(0.05/total_independent_tests[[1]])

# are we going to run mediation?
whether_mediate <- ifelse("MEDIATION" == "true", TRUE, FALSE)


```


## Manhattan plot

A genome-wide association study (GWAS) was performed by testing whether marker genotype differences can explain phenotypic variation. These tests correct for relatedness among individuals in the population using a genomic relatedness matrix (or "kinship matrix"). This anlaysis was performed with GCTA using two different kinship matrices: one constructed specifically with inbred model organisms in mind (INBRED) and one which is constructed from all markers except those on the chromosome of the tested marker ("leave-one-chromosome-out"; LOCO). The INBRED kinship matrix more heavily corrects for genetic stratification at the tested marker, while the LOCO kinship matrix does not, and may therefore increase power in certain contexts.

* **Every dot is a SNV marker.**

* **SNVs are colored if they pass the genome-wide corrected significance threshold:**
    * The horizontal solid line corresponds to stricter Bonferroni (BF) threshold which is based on the number of markers in the analysis.
    * The horizontal dash line corresponds to more permissive EIGEN threshold, which corrects for the number of independent markers in your data set. This threshold takes advantage of the extensive LD in *C. elegans* to limit the number of "unique" markers. (See Zdraljevic *et al*. 2019 (PMID: 30958264) for more)
    * If you selected a custom threshold, only this threshold is shown as a dotted line.

```{r, fig.height=6, fig.width=12}
# load processed mapping data. 
# Note that readr::read_delim will throw parsing errors and put NA in rows that contain actual values in strain/value/allele/var.exp so did not use it
# load inbred mapping reslts
processed_mapping_inbred <- read.delim(paste0("../../INBRED/Mapping/Processed/processed_",trait_name,"_AGGREGATE_mapping_inbred.tsv"), stringsAsFactors=FALSE) %>%
    dplyr::mutate(CHROM = factor(CHROM, levels = c("I","II","III","IV","V","X","MtDNA"))) %>%
    dplyr::select(-marker) %>%
    tidyr::unite("marker", CHROM, POS, sep = ":", remove = F) %>%
    dplyr::mutate(algorithm = "INBRED")

# load loco mapping results
processed_mapping_loco <- read.delim(paste0("../../LOCO/Mapping/Processed/processed_",trait_name,"_AGGREGATE_mapping_loco.tsv"), stringsAsFactors=FALSE) %>%
    dplyr::mutate(CHROM = factor(CHROM, levels = c("I","II","III","IV","V","X","MtDNA"))) %>%
    dplyr::select(-marker) %>%
    tidyr::unite("marker", CHROM, POS, sep = ":", remove = F) %>%
    dplyr::mutate(algorithm = "LOCO")

# combine both
processed_mapping <- rbind(processed_mapping_inbred, processed_mapping_loco)

# each trait has a separate processed_mapping file now. So the plotting function and loop is removed
# but do check there is only 1 trait and if not, issue warning:
num_traits = length(unique(dplyr::select(processed_mapping_inbred,trait)))

if(num_traits > 1){
    print("WARNING: More than 1 trait in processed_mapping table. Only the first one will be plotted.")
}

# do we have mito mapping?
mito_check <- processed_mapping %>%
    na.omit()

## MANHATTAN PLOTS ##
for.plot <- processed_mapping %>%
  dplyr::mutate(CHROM = as.factor(CHROM)) %>%
    {
        if(!("MtDNA" %in% mito_check$CHROM)) dplyr::filter(., CHROM != "MtDNA") else .
    }
BF <- processed_mapping %>% 
    dplyr::group_by(trait, algorithm) %>% 
    dplyr::filter(log10p != 0) %>% 
    dplyr::distinct(marker, log10p) %>%
    dplyr::mutate(BF = -log10(0.05/sum(log10p > 0, na.rm = T))) %>%
    dplyr::ungroup() %>%
    dplyr::select(BF) %>%
    unique(.) %>%
    dplyr::slice(1) %>% # BF can be slightly different between loco and inbred... but just plot one (5.46 v 5.47...)
    as.numeric()

EIGEN <- independent_test_cutoff
BF.frame <- processed_mapping %>%
  dplyr::select(trait) %>%
  dplyr::filter(!duplicated(trait)) %>%
  dplyr::mutate(BF = BF, EIGEN  = EIGEN, user = unique(processed_mapping$BF)[1])

# if user selected a different threshold, use that, otherwise plot BF and EIGEN
if(BF.frame$user %in% c(BF.frame$BF, BF.frame$EIGEN)) {
  for.plot.ann <- for.plot %>%
    dplyr::mutate(sig = case_when(log10p > BF.frame$BF ~ "BF",
                                  log10p > BF.frame$EIGEN ~ "EIGEN",
                                  TRUE ~ "NONSIG"))
  
  sig.colors <- c("red","#EE4266", "black")
  names(sig.colors) <- c("BF","EIGEN", "NONSIG")
} else {
  for.plot.ann <- for.plot %>%
    dplyr::mutate(sig = case_when(log10p > BF.frame$user ~ "user",
                                  TRUE ~ "NONSIG"))
  
  sig.colors <- c("red", "black")
  names(sig.colors) <- c("user", "NONSIG")
}

test <- BF.frame %>%
    tidyr::pivot_longer(BF:user) %>%
    dplyr::distinct() %>%
    dplyr::filter(name %in% names(sig.colors))

# are we plotting mito or no?
if("MtDNA" %in% unique(for.plot.ann$CHROM)) {
    facet_scales <- "fixed"
} else {
    facet_scales <- "free"
}


man.plot <- ggplot() + 
  theme_bw() + 
  geom_point(data = for.plot.ann, 
             mapping = aes(x = POS/1000000, 
                           y = log10p,
                           colour = sig,
                           alpha = sig)) +
  scale_alpha_manual(values = c("BF" = 1, "EIGEN" = 1, "user" = 1, "NONSIG" = 0.25)) +
  scale_colour_manual(values = sig.colors) + 
  scale_x_continuous(expand = c(0, 0), breaks = c(5, 10, 15, 20)) +
  geom_hline(data = test, aes(yintercept = value, linetype = name)) + 
  scale_linetype_manual(values = c("BF" = 1, "EIGEN" = 3, "user" = 2)) +
  labs(x = "Genomic position (Mb)",
       y = expression(-log[10](italic(p)))) +
  theme(legend.position = "none", 
        panel.grid = element_blank()) + 
  facet_grid(algorithm ~ CHROM, scales = "free_x", space = facet_scales) + 
  ggtitle(BF.frame$trait)


# save the plot to file. keep this code for user to create plots with their own styles
# ggsave(man.plot, filename = paste0(BF.frame$trait,"_manhattan.plot.png"), width = 8, height = 4)


# show plot in html output
man.plot

# plot all peaks
peaks <- processed_mapping %>%
    na.omit() %>%
    dplyr::select(CHROM, marker, log10p, trait, startPOS, peakPOS, endPOS, algorithm)%>%
    dplyr::distinct()  %>%
    dplyr::arrange(algorithm, marker)

DT::datatable(peaks)

```


## Genomic Inflation
The p-values calculated from each marker association test were compared to the theoretical distribution of p-values under the null hypothesis. This comparison is displayed for each chromosome in the quantile-quantile plots (Q-Q plots) below. The genomic inflation factor (λ_GC) estimates the inflation of observed p-values compared to a theoretical χ^2 [0.5,1]. Mappings producing genomic inflation factors greater than 1.25 may indicate some systematic bias, such as strong population stratification of phenotype values.

```{r, fig.height=6, fig.width=12}

expected.P.val.dists <- function(x,y){
  my.ps <- x %>%
    dplyr::select(marker, P)
  my.ps$exp.P <- (rank(my.ps$P, ties.method="first"))/(length(my.ps$P))
  
  x %>%
    dplyr::left_join(.,my.ps) %>%
    dplyr::mutate(trait = y)
}

# plot QQ-plot (expected v. actual)
# need to use distinct to get rid of the "QTL" processed part
exp_inbred <- expected.P.val.dists(processed_mapping_inbred %>% dplyr::distinct(marker, .keep_all = T), trait_name) %>%
  dplyr::mutate(algorithm = "INBRED") %>%
  dplyr::select(CHROM, P, exp.P, algorithm)
exp_loco <- expected.P.val.dists(processed_mapping_loco %>% dplyr::distinct(marker, .keep_all = T), trait_name) %>%
  dplyr::mutate(algorithm = "LOCO") %>%
  dplyr::select(CHROM, P, exp.P, algorithm)

qq_plot <- exp_inbred %>%
  dplyr::bind_rows(exp_loco) %>%
  dplyr::mutate(significant = ifelse(-log10(P) >= BF.frame$user, "yes", "no")) %>%
  dplyr::mutate(col = dplyr::case_when(algorithm == "INBRED" & significant == "yes" ~ "INBRED-significant",
                                       algorithm == "INBRED" & significant == "no" ~ "INBRED",
                                       algorithm == "LOCO" & significant == "yes" ~ "LOCO-significant",
                                       algorithm == "LOCO" & significant == "no" ~ "LOCO",
                                       TRUE ~ "NA")) %>%
    {
        if(!("MtDNA" %in% mito_check$CHROM)) dplyr::filter(., CHROM != "MtDNA") else .
    } %>%
  ggplot2::ggplot(.) +
  ggplot2::aes(x = -log10(exp.P), y = -log10(P), color = col) +
  ggplot2::geom_abline(slope = 1) +
  ggplot2::geom_point() +
  ggplot2::scale_color_manual(values = c("LOCO" = "darkslateblue", "INBRED" = "skyblue2", "LOCO-significant" = "red", "INBRED-significant" = "pink2")) +
  ggplot2::facet_grid(. ~ factor(CHROM)) +
  ggplot2::theme_bw() +
  ggplot2::labs(x = expression(-log[10](italic(p[expected]))), 
                y = expression(-log[10](italic(p[observed]))))

# save the plot to file. keep this code for user to create plots with their own styles
# ggsave(qq_plot, filename = paste0(trait_name,"_qq.plot.png"), width = 8, height = 4)


# show plot in html output
qq_plot

# calculate genomic inflation
calc.genomic.inflation <- function(x,y){
  gif.chisq <- qchisq(1-x$P,1)
  gif <- median(gif.chisq)/qchisq(0.5,1)
  trait <- y
  gif.df <- data.frame(gif, y)
  colnames(gif.df) <- c("GIF","trait")
  return(gif.df)
}

```

**The genomic inflation factor is `r calc.genomic.inflation(exp_inbred, trait_name)$GIF[1]` for the INBRED mapping and `r calc.genomic.inflation(exp_loco, trait_name)$GIF[1]` for the LOCO mapping**

*The following sections of the report are shown for mappings performed using both the INBRED and LOCO kinship matrix construction approaches. It is recommended you choose one set of results based on the previous diagnostic plots. These results may vary between different traits.*

---

## INBRED

This is the default kinship matrix construction approach, designed for inbred model organisms (See https://yanglab.westlake.edu.cn/software/gcta/#MakingaGRM for more info).

```{r}

inbred <- NULL

if(nrow(peaks %>% dplyr::filter(algorithm == "INBRED")) > 0) {
    run_inbred <- TRUE
} else {
    run_inbred <- FALSE
}


```

```{r, eval = run_inbred}

alg <- "INBRED"
inbred <- knitr::knit_child('NemaScan_Report_algorithm_template.Rmd')

```

`r if(!is.null(inbred)) { paste(knitr::knit(text = inbred), collapse = '\\n') }`

```{r, eval = !run_inbred}

print("No QTL were identified with the INBRED algorithm.")

```

<br>

---

## LOCO

LOCO may provide increased power to detect QTL because it does not correct for relatedness (or stratification) on the chromosome of each tested marker, sometimes providing higher power to detect linked QTL or QTL within divergent regions. However, this higher power also comes with higher false discovery rates. For more info, check out Widmayer *et al*. 2021 (https://www.biorxiv.org/content/10.1101/2021.09.09.459688v1).

```{r}

loco <- NULL

if(nrow(peaks %>% dplyr::filter(algorithm == "LOCO")) > 0) {
    run_loco <- TRUE
} else {
    run_loco <- FALSE
}


```

```{r, eval = run_loco}

alg <- "LOCO"
loco <- knitr::knit_child('NemaScan_Report_algorithm_template.Rmd')


```

`r if(!is.null(loco)) { paste(knitr::knit(text = loco), collapse = '\\n') }`

```{r, eval = !run_loco}

print("No QTL were identified with the LOCO algorithm.")

```

## Please kindly cite the following publications
* Widmayer SJ, Evans KS, Zdraljevic S, and Andersen EC (2021) Evaluating the power and limitations of genome-wide association mapping in C. elegans. bioRxiv 2021.09.09.459688.
* Lee D, Zdraljevic S, Stevens L, Wang Y, Tanny RE, Crombie TA, Cook DE, Webster AK, Chirakar R, Baugh LR, Sterken M, Braendle C, Felix M-A, Rockman MV, and Andersen EC (2020) Balancing selection maintains ancient genetic diversity in C. elegans. Nature Ecology and Evolution, 2021 Apr 5; DOI: 10.1038/s41559-021-01435-x.
* Zdraljevic S, Fox BW, Strand C, Panda O, Tenjo-Castano FJ, Brady SC, Crombie TA, Doench JG,  Schroeder FC, and Andersen EC (2019) Natural variation in arsenic toxicity is explained by differences in branched chain amino acid catabolism eLife, Apr 8;8: e40260.
* Cook DE, Zdraljevic S, Roberts JP, Andersen EC (2016) CeNDR, the Caenorhabditis elegans Natural Diversity Resource. Nucleic Acids Research, Jan 4; 45(D1):D650-D657.


<br>
<br>

```{r}
sessionInfo()
```
