---
title: "NemaScan Report for TRAIT_NAME_HOLDER"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_depth: 5
    toc_float:
      collapsed: true
      smooth_scroll: true
---

<br>

## Usage

* This R Markdown file imports data tables within Analysis_Results-Date folder and will knit without any modification.
<br>
* ggsave code is kept for users to make customized plots. 
<br>
<br>
<br>


```{r global_options, include=FALSE}
knitr::opts_chunk$set(echo=FALSE, warning=FALSE, message=FALSE)
options(knitr.duplicate.label = 'allow')
knitr::opts_knit$set(progress = FALSE)
```



```{r include=FALSE}
library.path <- .libPaths()
library("tidyverse", lib.loc = library.path)
library("plotly", lib.loc = library.path)
library("DT", lib.loc = library.path)
library("ggbeeswarm", lib.loc = library.path)
# library("ComplexHeatmap", lib.loc = library.path)
library("knitr", lib.loc = library.path)
library("ggrepel", lib.loc = library.path)
library("ggnewscale", lib.loc = library.path)
library("cowplot", lib.loc = library.path)

```

## Strain Issues

All strain names were converted to the corresponding isotype name, which can be looked up here: https://www.elegansvariation.org/strain/strain_data.tsv. If you submitted replicate data, replicates for a given isotype were averaged to one mean value.

```{r}

strain_issues <- readLines("Phenotypes/strain_issues.txt")

new <- grep(strain_issues, "WARNING:", value = T)
if(length(new) == 0) {
  print("No strain issues to report")
} else {
  print(new)
}

#no_issues <- c('"[1] \"Strain issues: (if empty, no strain issues were found)\""', '"Warning message:"', '"Grouping rowwise data frame strips rowwise nature "')

# check to see if there were any issues:
#if(identical(strain_issues, no_issues)) {
#    "No strain issues to report!"
#} else {
#    strain_issues[2:(length(strain_issues)-2)]
#}

```


```{r}
# load trait name
# in nextflow use sed to edit this field and make a copy of the .rmd for each trait
trait_name <- "TRAIT_NAME_HOLDER"   

# load independent tests result
total_independent_tests <- read.table("Genotype_Matrix/total_independent_tests.txt", quote="\"", comment.char="", stringsAsFactors=FALSE)

independent_test_cutoff <- -log10(0.05/total_independent_tests[[1]])

```


## Manhatton plot
* #### Every dot is a SNV marker.

* #### SNVs are colored red if they pass the genome-wide corrected significance threshold:
    * If Bonferroni-corrected significance (BF) threshold were used: the horizontal solid line corresponds to BF threshold, and the horizontal dash line corresponds to EIGEN threshold (see below). If user supplied threshold exists, it will replace the BF threshold.
    * If Eigen-decomposition significance threshold were used, the horizontal solid line overlaps with the dash line and corresponds to the EIGEN threshold, which corrects for the number of independent markers in your data set.

* #### Light blue regions are QTL confidence intervals. These regions have separate report pages with links below.


```{r, fig.height=4.5, fig.width=12}
# load processed mapping data. 
# Note that readr::read_delim will throw parsing errors and put NA in rows that contain actual values in strain/value/allele/var.exp so did not use it
processed_mapping <- read.delim(paste0("Mapping/Processed/processed_",trait_name,"_AGGREGATE_mapping.tsv"), stringsAsFactors=FALSE) %>%
    dplyr::mutate(CHROM = factor(CHROM, levels = c("I","II","III","IV","V","X","MtDNA"))) %>%
    dplyr::select(-marker) %>%
    tidyr::unite("marker", CHROM, POS, sep = ":", remove = F)

# each trait has a separate processed_mapping file now. So the plotting function and loop is removed
# but do check there is only 1 trait and if not, issue warning:
num_traits = length(unique(dplyr::select(processed_mapping,trait)))

if(num_traits > 1){
    print("WARNING: More than 1 trait in processed_mapping table. Only the first one will be plotted.")
}

## MANHATTAN PLOTS ##
BF.frame <- processed_mapping %>%
    dplyr::select(trait, BF) %>%
    dplyr::filter(!duplicated(trait))
for.plot <- processed_mapping %>%
    dplyr::mutate(CHROM = as.factor(CHROM)) %>%
    dplyr::filter(CHROM != "MtDNA") %>%
    dplyr::mutate(algorithm = as.factor(algorithm))
  
man.plot <- ggplot2::ggplot() + 
    ggplot2::theme_bw() + 
    ggplot2::geom_point(data = for.plot[which(for.plot$aboveBF == 1),], 
                        mapping = aes(x = POS/1000000, 
                                      y = log10p), shape = 21, fill = "red") +
    ggplot2::geom_point(data = for.plot[which(for.plot$aboveBF == 0),], 
                        mapping = aes(x = POS/1000000, 
                                      y = log10p), 
                        alpha = 0.25) +
    ggplot2::scale_y_continuous(expand = c(0,0), limits = c(0,max(for.plot$log10p + 1))) +
    ggplot2::geom_hline(data = BF.frame, aes(yintercept = BF), linetype = 2) + 
    ggplot2::labs(x = "Genomic position (Mb)",
                  y = expression(-log[10](italic(p)))) +
    ggplot2::theme(legend.position = "bottom", 
                   panel.grid = element_blank()) + 
    ggplot2::facet_grid(. ~ CHROM, scales = "free_x", space = "free") + 
    ggplot2::ggtitle(BF.frame$trait)

# save the plot to file. keep this code for user to create plots with their own styles
# ggsave(man.plot, filename = paste0(BF.frame$trait,"_manhattan.plot.png"), width = 8, height = 4)


# show plot in html output
man.plot
```



```{r}
# read in QTL regions for all traits. Note that QTL_peaks.tsv is generated even if there is no significant regions.
# based on number of regions & width of regions to decide whether generate pxg and fine mapping plots

# what happens to this file if no QTL?
if (file.info("Mapping/Processed/QTL_peaks.tsv")$size != 0) {

    whether_print_peaks <- TRUE
        
    QTL_peaks <- read.delim("Mapping/Processed/QTL_peaks.tsv", header=TRUE, stringsAsFactors=FALSE) %>%
        dplyr::filter(trait == trait_name)

    # plot pxg?
    if(nrow(QTL_peaks == 0)) {
        whether_plot_pxg <- FALSE
    } else {
        whether_plot_pxg <- TRUE
    }
    
    # don't have LD yet
    if (nrow(QTL_peaks) < 2) {
        whether_plot_LD <- FALSE
    } else { whether_plot_LD <- TRUE }
    
    if (nrow(QTL_peaks) == 0) {
        print("No significant QTL region for this trait")
        whether_fine_map <- FALSE # skip related chunks
        whether_plot_LD <- FALSE
    } else {

        # QTL_peaks <- QTL_peaks %>%
        #     dplyr::mutate(Region_width = endPOS - startPOS) %>% 
        #     dplyr::filter(Region_width < 5000000)
        
        if (nrow(QTL_peaks)==0) {
            print("Significant QTL regions are too wide for plotting")
            whether_fine_map <- FALSE # skip related chunks
            
        } else {
            whether_fine_map <- TRUE # run related chunks
            
        }
    }
    
} else { 
    print("No significant peak for this trait")
    whether_print_peaks <- FALSE
    whether_plot_LD <- FALSE
    whether_fine_map <- FALSE
}

```


```{r, eval = whether_print_peaks}
# print out the table to html
# this doesn't print within if{}
# QTL_peaks was filtered in previous chunk, so reload
QTL_peaks_original <- read.delim("Mapping/Processed/QTL_peaks.tsv", header=TRUE, stringsAsFactors=FALSE) %>% 
        dplyr::filter(trait == trait_name)

DT::datatable(QTL_peaks_original)

```

<br>
<br>

## Phenotype by Genotype Split
```{r, eval=whether_plot_pxg}


## SWEPTNESS & EFFECTS SUMMARY ##
# proc.sweeps <- sweeps %>%
#     dplyr::select(c(isotype,contains("hapshare")))
# colnames(proc.sweeps) <- gsub(colnames(proc.sweeps),pattern = "_hapshare", replacement = "")
# sweep.chrom.pivot <- proc.sweeps %>%
#     tidyr::pivot_longer(cols = -isotype, names_to = "CHROM", values_to = "sweep.share") %>%
#     dplyr::rename(strain = isotype)


nested.pxg.dat <- processed_mapping %>%
    dplyr::filter(!is.na(peak_id)) %>%
    dplyr::select(CHROM, marker, trait, startPOS, peakPOS, endPOS, AF1, value, strain, allele, peak_id) %>%
    dplyr::distinct() %>%
    dplyr::mutate(startPOS = startPOS/1000000,
                  peakPOS = peakPOS/1000000,
                  endPOS = endPOS/1000000) %>%
    # dplyr::left_join(.,sweep.chrom.pivot) %>% # don't have this file yet
    # dplyr::group_by(trait, peak_id) %>%
    # dplyr::recode(allele, "-1" = "REF", "1" = "ALT") %>%
    dplyr::mutate(allele = dplyr::case_when(allele == "-1" ~ "REF",
                                            allele == "1" ~ "ALT",
                                            TRUE ~ "NA"),
                  allele = factor(allele, levels = c("REF", "ALT")))
    
CB.N2.code <- nested.pxg.dat %>%
    dplyr::select(strain, allele) %>%
    dplyr::distinct() %>%
    dplyr::filter(strain %in% c("CB4856","N2","PD1074")) %>%
    droplevels()
    
if(length(levels(CB.N2.code$allele)) < 2){
  pal <- c("#726E75","#720E07")
} else {
  pal <- c("#FFA500","#0000ff")
}

strains.of.interest <- c("PD1074", "N2", "CB4856", "RC301", "MY16", 
                             "ECA396", "ECA36", "XZ1516", "ECA248", "AB1", 
                             "CB4507", "CB4858", "CB4855", "CB4852", "MY1", 
                             "JU319", "JU345", "JU400", "PB306", "PX174", "PX179")

# plot_title <- glue::glue("{unique(nested.pxg.dat$trait)}:CHR{unique(nested.pxg.dat$CHROM)}:{round(unique(nested.pxg.dat$peakPOS), digits = 2)}MB")

plot <- nested.pxg.dat %>%
    dplyr::filter(allele != "NA" | !is.na(allele)) %>%
    dplyr::mutate(SOI = strain %in% strains.of.interest,
                    SOI.2 = if_else(SOI == TRUE, true = strain, false = "")) %>%
    droplevels() %>%
    dplyr::arrange(SOI.2) %>%
    ggplot2::ggplot(mapping = aes(x = allele, y = value, text = SOI.2)) +
    ggplot2::theme_bw(base_size = 12) +
    ggplot2::geom_violin(aes(fill = allele), alpha = 0.8, scale = "count", draw_quantiles = c(0.25, 0.5, 0.75)) +
    ggplot2::scale_fill_manual(values = pal, guide = FALSE) +
    ggnewscale::new_scale("fill") +
    ggplot2::geom_point(aes(fill = SOI), position = ggbeeswarm::position_beeswarm(), size = 1.5, shape = 21) +
    # geom_point(aes(colour = sweep.share*100), size = 1.1, position = pos) +
    ggplot2::scale_fill_manual(values = c("#9297C4","#D33E43"), guide = FALSE) +
    # scale_colour_gradient(low = "black", high = "violetred", name = "Selective Sweep (% Chromosome)") +
    ggrepel::geom_text_repel(aes(label = SOI.2),
                      colour = "black", position = ggbeeswarm::position_beeswarm()) +
    ggplot2::theme(legend.position = "bottom") +
    ggplot2::labs(y = "Trait Value",
           x = "Genotype") +
    ggplot2::facet_grid(~marker)

print(plot)

# interactive plots - doesn't work well with facet?
# plotly::ggplotly(plot, tooltip = c("plotly_label", "y")) %>%
#     layout(autosize = F, width = 400, height = 400)

# ggsave(glue::glue("{plot_title}_effect.plot.png"), height = 5, width = 5)


```

## Linkage disequilibrium (LD) between QTL
```{r, eval = whether_plot_LD}

# this code is not very clean for a markdown for other people to follow... maybe we should compute LD first then output results?
compute.LD <- function(QTL, trait){
  if(length(unique(QTL$peak_id)) <= 1){
    data.frame(unique(QTL$marker),unique(QTL$marker),NA, trait) %>%
      `colnames<-`(c("QTL1","QTL2","r2","trait"))
  } else {
    QTLcombos <- data.frame(t(combn(x = unique(QTL$peak_id), m = 2))) %>%
      `colnames<-`(c("QTL1","QTL2"))
    
    LD <- list()
    for(i in 1:nrow(QTLcombos)){
      print(i)
      peaks.of.interest <- as.numeric(QTLcombos[i,])
      haps <- QTL %>%
        dplyr::select(strain, allele, peak_id) %>%
        dplyr::filter(peak_id %in% peaks.of.interest) %>%
        dplyr::mutate(allele = if_else(allele == -1, true = "REF", false = "ALT")) %>%
        dplyr::mutate(peak_id = as.factor(peak_id)) %>%
        tidyr::pivot_wider(names_from = peak_id, values_from = allele) %>%
        `colnames<-`(c("strain","A","B")) %>%
        tidyr::unite("hap", c(A,B), sep = "_", remove = F)
      
      P <- haps %>%
        dplyr::group_by(hap) %>%
        dplyr::summarise(n()/nrow(haps)) %>%
        `colnames<-`(c("P","freq")) %>%
        tidyr::pivot_wider(names_from = P, values_from = freq)
      
      n <- QTL %>%
        dplyr::select(peak_id) %>%
        dplyr::mutate(peak_id = as.factor(peak_id)) %>%
        dplyr::group_by(peak_id) %>%
        dplyr::summarise(n()) %>%
        `colnames<-`(c("peak_id","total"))
      
      p <- QTL %>%
        dplyr::select(strain, allele, peak_id) %>%
        dplyr::filter(peak_id %in% peaks.of.interest) %>%
        dplyr::mutate(allele = if_else(allele == -1, true = "REF", false = "ALT")) %>%
        dplyr::mutate(peak_id = as.factor(peak_id)) %>%
        dplyr::group_by(allele, peak_id) %>%
        dplyr::summarise(n()) %>%
        `colnames<-`(c("p","peak_id","n")) %>%
        dplyr::left_join(.,n) %>%
        dplyr::mutate(freq = n/total) %>%
        tidyr::unite("allele", c(p,peak_id), sep = "_") %>%
        dplyr::ungroup() %>%
        dplyr::select(allele, freq) %>%
        tidyr::pivot_wider(names_from = allele, values_from = freq)
      
      pApB <- p %>%
        dplyr::select(contains("REF")) %>%
        c(.) %>%
        unlist() %>%
        prod()
      
      pApBpapb <- p %>%
        unlist() %>%
        prod()
      
      D.AB <- P$REF_REF - pApB
      r2 <- (D.AB^2)/pApBpapb
      LD[[i]] <- QTL %>%
        dplyr::select(marker, peak_id) %>%
        dplyr::filter(peak_id %in% peaks.of.interest) %>%
        dplyr::distinct()  %>%
        tidyr::pivot_wider(names_from = peak_id, values_from = marker) %>%
        `colnames<-`(c("QTL1","QTL2")) %>%
        dplyr::mutate(r2 = r2,
                      trait = trait)
    }
    Reduce(rbind, LD)
  }
}

## LD PLOTS ##
nested.QTL <- processed_mapping %>%
  as.data.frame() %>%
  dplyr::filter(!is.na(peak_id), !is.na(allele)) %>%
  dplyr::select(CHROM, marker, trait, algorithm, AF1, value, strain, allele, peak_id) %>%
  dplyr::distinct() %>%
  dplyr::group_by(trait) %>%
  tidyr::nest()

trait.LD <- purrr::pmap(.l = list(nested.QTL$data, 
                                  nested.QTL$trait), 
            .f = compute.LD) %>%
  Reduce(rbind,.)

options(scipen = 999)
if(!is.null(trait.LD)){
  check <- trait.LD %>%
    dplyr::filter(!is.na(r2)) %>%
    nrow()
  if(check != 0){
    LD.plot <- trait.LD %>%
      dplyr::filter(!is.na(r2)) %>%
      ggplot(., mapping = aes(x = QTL1, y = QTL2)) + 
      theme_classic() +
      geom_tile(aes(fill = r2),colour = "black", size = 3) + 
      geom_text(aes(label = round(r2, 4))) + 
      scale_fill_gradient(low="darkgreen", high="red", limits = c(0, 1), name = expression(italic(r^2))) + 
      theme(axis.title = element_blank(),
            axis.text = element_text(colour = "black")) + 
      labs(title = paste0("Linkage Disequilibrium: ",unique(trait.LD$trait)))
    
    print(LD.plot)
    # ggsave(LD.plot, filename = paste0(unique(trait.LD$trait),"_LD.plot.png"), width = 7, height = 7)
  }
}

```

```{r, eval=!whether_plot_LD}
print("Not enough QTL regions to calculate linkage")
```

<br>
<br>

## Divergent region, haplotype, and fine mapping for each significant QTL region 
Only for regions less than 5 Mb in size

```{r, eval = whether_fine_map}

out <- NULL
for (i in seq_len(nrow(QTL_peaks))) { 
    QTL_chrom <- QTL_peaks$CHROM[i]
    QTL_start <- QTL_peaks$startPOS[i]
    QTL_end <- QTL_peaks$endPOS[i]
    QTL_Peak <- QTL_peaks$peakPOS[i]
    region_name <- glue::glue("{QTL_chrom}:{QTL_start}-{QTL_end}")
    
    out <- c(out, knitr::knit_child('NemaScan_Report_region_template.Rmd'))
}

```

`r paste(knitr::knit(text = out), collapse = '\n')`

<br>
<br>

```{r, eval=!whether_fine_map}
print("No significant QTL regions to plot")
```

<br>
<br>


## Please kindly cite the following publications
* Cook, D.E., Zdraljevic, S., Roberts, J.P., and Andersen, E.C. (2016). CeNDR, the Caenorhabditis elegans natural diversity resource. Nucleic Acids Res.
* Zdraljevic, S., Fox, B.W., Strand, C., Panda, O., Tenjo, F.J., Brady, S.C., Crombie, T.A., Doench, J.G., Schroeder, F.C., and Andersen, E.C. (2019). Natural variation in C. elegans arsenic toxicity is explained by differences in branched chain amino acid metabolism. Elife 8.
* Lee, D., Zdraljevic, S., Stevens, L., Wang, Y., Tanny, R.E., Crombie, T.A., Cook, D.E., Webster, A.K., Chirakar, R., Ryan Baugh, L., et al. (2020). Balancing selection maintains ancient genetic diversity in C. elegans.
* haplotype paper

<br>
<br>

```{r}
sessionInfo()
```
